rfd/0056-buildbox-image-handling.md

```diff
@@ -0,0 +1,90 @@
+    the use of privileged containers (e.g. require Docker-In-Docker), as this 
+    is a security threat.
+
+ 6. Creating an updated buildbox image once changes have been approved should 
+    be automatic (or at least a single push-button operation)
+
+ 7. Where multiple build environments are required, they should all be 
+    produced by the same process.
+
+### The current process
+
+Currently the buildbox images are constructed by Drone in response to a commits
+being pushed onto `gravitational/teleport:master`. The drone pipeline constructs 
+updated buildbox images from the latest source in `build.assets/Dockerfile` (and 
+it's siblings), tags them with the version of `go` installed on the image and 
+then pushed them to the repository at `quay.io`. These builds are then used for CI.
+
+For _releases_, the release pipelines in Drone construct a buildbox image at 
+_build time_ from the same source in `build.assets`, and then execute the build 
+inside it via Docker-in-Docker
+
+This process meets several of the goals set above:
+
+ * The buildbox for a release are constructed from source at build time, so the 
+   current process (somewhat - I'll explain why below) meets Goal #1.
+ * Changes to the buildbox are incorporated by the PR/code review process, so 
+   Goal #3 is met
+ * Images are automatically produced for both CI and Release, so Goal #4 is met
+ * Images for all of the different build targets are created in the same way, so
+   Goal #7 is met
+
+On the other hand:
+
+ * We do no validation of the components installed onto the buildbox image, so 
+   the every image created (including at build time for releases) is vulnerable
+   to downloaded artifacts being subverted.
+ * We tag _every_ revision of the buildbox with the current go version, meaning
+   that we cannot uniquely identify any revision of the buildbox image.
+ * Using the version of Go as the image tag means we have no mechanism to 
+   independently upgrade other software in the image, or identify which images 
+   contain what versions of, for example, Rust.
+ * Updating the buildbox image automatically forces and CI jobs using the 
+   buildbox to accept the new changes. This has recently caused issues where 
+   updating the buildbox `go1.17.2` image to build Teleport 9 broke CI for 
+   Teleport 8, as there is no way to differentiate the images they require.
+
+### Suggested changes
+
+* Independently versioned buildboxes, so we can deal with things like Go and 
+  Rust being updated independently
+* Buildbox source extracted from Teleport repo into it's own space
+* Images get build on a tag as per releases, after code review & CI as per
+  general software.
+* Images being pushed to a repository with immutable tags, so we have a guarantee
+  that the image is what we think it is
+* Everything (including releases) uses the same set of images
+* Enforce strong validation of components on the buildbox as far as possible
```

Comment:
Two thoughts about strong validation of components:

1) A pull through artifact cache is probably the state of the art.  We have a huge range of artifacts though, and pulling this in scope would be notable scope creep.  I've had good experiences with artifactory at previous shops, and I think @fheinecke was researching Package Cloud.
2) The next best thing is pinning SHAs for everything fetched from the internet.  This doesn't guarantee stuff will remain available, but it does guarantee the build will break if anything unexpected changes.  Nix takes this approach and it works well with a "fail forward" mindset -- rollback become difficult-to-impossible.

Reply:
Goal #1 for this is not to be surprised by a change after the fact. Caching dependencies is great, but that's a secondary consideration (at least to me)

Just for context, the main experience I have with this sort of thing is reliably building an OpenEmbedded distro, which has SHAs for _everything_ that gets installed. That's obviously overkill for this, especially if we plan to continue using Ubuntu as a base, where we just have to trust that their debian packages have not been molested. But that's the environment I'm starting with, and has shaped my perspective.

I have a mate who works at Octopus Deploy (who, incidentally worked on the OpenEmbedded project with me) - might be worth a chat.

Reply:
_**1 year later:**: We're now using Centos7 as a based. Same comments apply to the RedHat package archive, though.